src/common/tusb_fifo.c

```diff
@@ -173,55 +173,80 @@ static inline void _ff_push(tu_fifo_t* f, void const * data, uint16_t wRel)
   memcpy(f->buffer + (wRel * f->item_size), data, f->item_size);
 }
 
-static inline void _ff_push_copy_fct(void * dst, const void * src, uint16_t len, tu_fifo_copy_mode_t copy_mode)
-{
-  switch (copy_mode)
-  {
-    case TU_FIFO_COPY_INC:
-      memcpy(dst, src, len);
-      break;
-
-    case TU_FIFO_COPY_CST:
-      _tu_fifo_read_from_const_src_ptr(dst, src, len);
-      break;
-  }
-}
-
-static inline void _ff_pull_copy_fct(void * dst, const void * src, uint16_t len, tu_fifo_copy_mode_t copy_mode)
+// send n items to FIFO WITHOUT updating write pointer
+static void _ff_push_n(tu_fifo_t* f, void const * data, uint16_t n, uint16_t wRel, tu_fifo_copy_mode_t copy_mode)
 {
   switch (copy_mode)
   {
     case TU_FIFO_COPY_INC:
-      memcpy(dst, src, len);
+      if(n <= f->depth-wRel)
+      {
+        // Linear mode only
+        memcpy(f->buffer + (wRel * f->item_size), data, n*f->item_size);
+      }
+      else
+      {
+        // Wrap around
+        uint16_t nLin = f->depth - wRel;
+
+        // Write data to linear part of buffer
+        memcpy(f->buffer + (wRel * f->item_size), data, nLin*f->item_size);
+
+        // Write data wrapped around
+        memcpy(f->buffer, ((uint8_t const*) data) + nLin*f->item_size, (n - nLin) * f->item_size);
+      }
       break;
 
-    case TU_FIFO_COPY_CST:
-      _tu_fifo_write_to_const_dst_ptr(dst, src, len);
+    case TU_FIFO_COPY_CST_FULL_WORDS:   // Intended for hardware buffers from which it can be read word by word only
+      if(n <= f->depth-wRel)
+      {
+        // Linear mode only
+        _tu_fifo_read_from_const_src_ptr_in_full_words(f->buffer + (wRel * f->item_size), data, n*f->item_size);
+      }
+      else
+      {
+        // Wrap around case
+        uint16_t nLin = (f->depth - wRel) * f->item_size;
+        uint16_t nWrap = (n - nLin) * f->item_size;
+
+        uint8_t * dst_u8 = (uint8_t *)(f->buffer + (wRel * f->item_size));
+        volatile uint32_t * rx_fifo = (volatile uint32_t *) data;
+        CFG_TUSB_MEM_ALIGN uint32_t tmp;
+
+        // Write full words of linear part to buffer
+        uint16_t full_words = nLin >> 2;
+        for(uint16_t i = 0; i < full_words; i++) {
+          tmp = *rx_fifo;
+          memcpy(dst_u8, &tmp, 4);
```

Comment:
It should be better to do direct copy, there are many overhead inside `memcpy`

Reply:
Ok, but isn't memcpy optimized anyway since it has constant 4 bytes to copy?
I refactored the code for handling the wrapping case as there the number of bytes to copy is not constant.

Reply:
I see your point. Since `dst_u8` is not word-aligned the compiler is not always willing to optimize `memcpy` out, for example in IAR I have to use highest optimization or it just call `__aeabi_memcpy`. There is `__attribute__((__packed__))` for this kind of stuff.

You can do something like this:
```
typedef struct{
    int i;
} __attribute((__packed__)) unaligned_int_t;

unaligned_int_t* dst = (unaligned_int_t*)(f->buffer + (wRel * f->item_size));;
dst->i = *rx_fifo;
```

There is no need to add `CFG_TUSB_MEM_ALIGN` in front of `uint32_t` since it's always aligned.

Reply:
Fany ^^ But it works, i implemented it and pushed the changes. I hope the CI runs through ^^ Thanks for advice!

Reply:
@HiFiPhile I don't like this change, this complicate the code more than it should be. How much overhead we are talking about here. I am more concerned about the overhead in our time to process this command in our thought though.

Reply:
@hathach 
> this complicate the code more than it should be

 I don't see it complicate the code, it's a quite common way to do unaligned copy.

> How much overhead we are talking about here

Unless I set optimization to high (which cause problem in step by step debugging), IAR generate this code and it stays for 43 cycles inside `__aeabi_memcpy4` on my STM32L412.
```
        ADD      R0,SP,#+20     
        ADR.W    R1,?_0         
        MOVS     R2,#+16        
          CFI FunCall __aeabi_memcpy4
        BL       __aeabi_memcpy4
```

>  I am more concerned about the overhead in our time to process these line

I don't see this `struct unaligned_int_t` can generate any more overhead than memcpy ?

Reply:
You are right,  43 cycles for a word is indeed quite a lot considering we are doing hundreds of this. And a bit of googling, also suggests it is common to address this. It is my bad.  Though maybe we should declare it as global typedef for these to reduce duplication (could do this). btw do you think having these align would work instead of using packed struct ? 

```C
typedef uint64_t unaligned_uint64_t __attribute__ ((aligned(1)));
typedef uint32_t unaligned_uint32_t __attribute__ ((aligned(1)));
typedef uint16_t unaligned_uint16_t __attribute__ ((aligned(1)));
```

Since you are in there, could you also compare the above unaligned assignment compared to the good old shift + or. I am just curious
 
```
( b1 << 24) | ( b2 << 16) | (b3 << 8) | b4
``

Reply:
Unfortunately it's not work without struct, gcc simply ignored `__attribute__ ((aligned(1)))`, I've also tried IAR (IAR now support most gcc syntaxes).

In IAR I can simply do `dst = *(__packed uint32_t*)(f->buffer + (wRel * f->item_size));` but there is no packed pointer in gcc.

If src is not `volatile` shift + or and packed struct generate nearly the same code, on ARMv6M:
```
#Shift + Or
	ldr	r0, .LCPI0_0
	ldrb	r1, [r0, #5]
	lsls	r1, r1, #24
	ldrb	r2, [r0, #6]
	lsls	r2, r2, #16
	adds	r1, r2, r1
	ldrb	r2, [r0, #7]
	lsls	r2, r2, #8
	adds	r1, r1, r2
	ldrb	r0, [r0, #8]
	adds	r0, r1, r0
	ldr	r1, .LCPI0_1
	str	r0, [r1]
    
    
#Packed struct    
	ldr	r0, .LCPI0_0
	ldrb	r1, [r0, #5]
	ldrb	r2, [r0, #6]
	lsls	r2, r2, #8
	adds	r1, r2, r1
	ldrb	r2, [r0, #7]
	ldrb	r0, [r0, #8]
	lsls	r0, r0, #8
	adds	r0, r0, r2
	lsls	r0, r0, #16
	adds	r0, r0, r1
	ldr	r1, .LCPI0_1
	str	r0, [r1]
```
On ARMv7M a `LDR.W` is corretly generated.

If src is `volatile`  shift + or generate above code.

Reply:
Thank you for your extra information. I have learned something new (still learn new thing with C every now and then).  Using memcpy to replace shift-or is my bad. We have inline helper, maybe we can put them into  usage here
https://github.com/hathach/tinyusb/blob/master/src/common/tusb_common.h#L85

 I am still prefer doing thing more explicit than leaving it to compiler, especially if it is not too obvious. I really need to wrap up the review and merge this PR asap :sweat_smile:

Reply:
Yes we can use the inline version.

We only intended to support 32bit little endian systems I think?

Reply:
@HiFiPhile ah you are right, there is also native mcu endian matter as well, maybe it is better this way. I will go through fifo.c again later on to see what is best suited for. Thank you for bringing this up :)

Reply:
For info we can do it like zephyr for GCC, https://github.com/zephyrproject-rtos/zephyr/blob/65ac2f6d1bf9570da68b2da30e769aab2f39b7c2/include/toolchain/gcc.h#L102

I will take care of IAR :)